### PR TITLE
Login/authentication enhancements

### PR DESCRIPTION
**Pull request by Wauplin**
This PR do various things related to authentication. I started it to support Google Colab secrets and ended up rewriting some docs and deprecating some legacy code.

**List of changes:**
1. New authentication section in the quick-start guide. Especially:
    - document how to login via CLI
    - encourage login with env variable for Google Colab and Spaces
    - discourage hardcoded tokens
    - warnings about token permission and leaks
    - => this section is in the quick-start guide. I hesitated to have a dedicated guide/page for it but I don't think it's worth it (not much content + we would still have some "authentication" content in the quick start so let's keep everything together)
2. **Automatically read `HF_TOKEN` secret when running in a colab notebook**! :fire: 
3. `logout` now raises an error if token is deleted but user was logged in via environment variable / secret value. I think this is important otherwise a user can think they are completely logged out of a machine when it's not the case.
4. Started to deprecate `HfFolder` in favor of `get_token`, `login` and `logout`.
    - `huggingface_hub.get_token` feels much more intuitive `huggingface_hub.Hffolder.get_token`
    - `login` in favor of `HfFolder.save_token` => saves token but also checks token validity. I don't think an external lib should save the token without checks.
    - `logout` in favor of `HfFolder.delete_token`  => deletes token but also checks if the user is really logged out (i.e. no env variable)
    - => for now, `HfFolder` is not officially deprecated. I will first release this, then adapt `transformers`/`datasets`/`gradio` codebase and then deprecate it officially in the next release (and most likely remove in `v1.0`).
5. ignore token if empty string (`""`) and consider as None
6. adapt tests

**TODO:**
- [ ] check documentation in other languages
- [ ] fix remaining tests

image

### CODE REVIEW

File: docs/source/en/quick-start.md [hunk 1]
```diff
@@ -51,45 +51,67 @@ full-length hash instead of the shorter 7-character commit hash:
...
-If you want to handle several accounts in the same script, you can provide your token when calling each method. This
-is also useful if you don't want to store any token on your machine.
+Finally, it is also possible to authenticate programmatically by passing your token to any method that accepts `token` as parameter. This is usually discouraged except in an environment where you don't want to store your token permanently or if you need to handle several tokens at once.
 
 <Tip warning={true}>
 
-Once you are logged in, all requests to the Hub -even methods that don't necessarily require authentication- will use your
-access token by default. If you want to disable implicit use of your token, you should set the
-`HF_HUB_DISABLE_IMPLICIT_TOKEN` environment variable.
+Please be careful when passing tokens as parameter. It is always best practice to load the token from a secure vault instead of hardcoding it in your codebase or notebook. Hardcoded tokens present a major risk of leak if you share your code inadvertently.
```
Comment (LysandreJik) [inline docs/source/en/quick-start.md:114]: Yes! We'll probably need to do a few updates in the libraries that depend on it as we were encouraging this for now

**Review by LysandreJik — APPROVED** [review]
Yes, agree with the refactor! Nice to see the updates. I agree that moving away from `token=xxx` as a default within code is better for security; we'll need to update downstream documentation as well as I'm pretty sure we have a lot of references to `token` in `transformers`/`diffusers`.

LGTM!

**Review by stevhliu — APPROVED** [review]
LGTM!

I did a quick search of the Diffusers/Transformers repos with `repo:huggingface/transformers token= language:Markdown` and the only docs I found that reference `token=...` is the NLLB doc [here](https://huggingface.co/docs/transformers/v4.35.2/en/model_doc/nllb#generating-from-any-other-language-than-english).

**Comment by julien-c** [comment]
> I agree that moving away from `token=xxx` as a default within code is better for security

Honestly, let's not go overboard with this, given anyways both on the Hub and on GH, any uploaded HF token will soon be auto-invalidated

So honestly i'm not sure it's a huge issue

File: docs/source/en/quick-start.md [hunk 2]
```diff
@@ -51,45 +51,67 @@ full-length hash instead of the shorter 7-character commit hash:
...
 the Hub: download private repos, upload files, create PRs,...
 [Create an account](https://huggingface.co/join) if you don't already have one, and then sign in
 to get your [User Access Token](https://huggingface.co/docs/hub/security-tokens) from
 your [Settings page](https://huggingface.co/settings/tokens). The User Access Token is
 used to authenticate your identity to the Hub.
 
-Once you have your User Access Token, run the following command in your terminal:
+<Tip>
+
+Tokens can have `read` or `write` permissions. Make sure to have a `write` access token if you want to create or edit a repository. Otherwise, it's best to generate a `read` token to reduce risks in case you leak your token inadvertently.
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply docs/source/en/quick-start.md:67]: Ahah, yes that sounds better :smile:

File: docs/source/en/quick-start.md [hunk 3]
```diff
@@ -51,45 +51,67 @@ full-length hash instead of the shorter 7-character commit hash:
...
+
+Another possibility to authenticate is to set the environment variable `HF_TOKEN` with your access token. This is especially useful in Spaces where you can set `HF_TOKEN` as a [Space secret](https://huggingface.co/docs/hub/spaces-overview#managing-secrets).
+
+<Tip>
+
+**NEW:** Google Colaboratory let you [define private keys](https://twitter.com/GoogleColab/status/1719798406195867814) for your notebooks. Define a `HF_TOKEN` secret to be automatically authenticated!
+
+</Tip>
+
+Authentication via an environment variable or a secret has priority over the token stored on your machine.
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply docs/source/en/quick-start.md:106]: What I meant is if a token is stored on the machine (via `huggingface-cli login` AND the environment variable is set, then the environment variable is used (so "it has priority").

**Comment by Wauplin** [comment]
Thanks both for the reviews! I addressed all of the comments I think. 
I have updated the Google Colab integration to handle errors nicely. The logic is:
- if user has defined a `HF_TOKEN` secret and granted access to the notebook => use it => normal use case
- if user has defined a `HF_TOKEN` secret but did not grant access to the notebook, a pop-up will ask permission
   - if permission is granted by the user => use the token. Next calls will not need to re-request the access.
   - if permission is NOT granted =>  warn the user they'll not be authenticated + set flag to ensure we do not ask again + do not fail (returns `None` as token)
- if `HF_TOKEN` is not defined in the user secrets => warn to say "hey it's recommended to define a token" + return `None` 
- if anything else happen => warn "You are not authenticated and we don't know why. Please report on GitHub" + return `None`

**Comment by Wauplin** [comment]
@LysandreJik could you re-review logic for `_get_token_from_google_colab` in `src/huggingface_hub/_login.py` please? (see https://github.com/huggingface/huggingface_hub/pull/1895#issuecomment-1850356442) It's the only piece of real logic that I have updated. Apart from that, tests look ok and we should be good to merge.

**Review by LysandreJik — APPROVED** [review]
Impressive PR @Wauplin! This looks good to me. Thanks for working on this.

**Comment by Wauplin** [comment]
Thanks for the last review @LysandreJik, let's merge this! :)

**Pull request merged**